apps/chat/tests/test_scheduled_messages.py

```diff
@@ -0,0 +1,320 @@
+        # Set the DB time to now
+
+        # DB is now behind the next trigger date
+        poll_scheduled_messages()
+        test_scheduled_message_attrs(
+            expected_next_trigger_date=current_time + delta,
+            expected_last_triggered_at=None,
+            expected_total_triggers=0,
+            expected_resolved=False,
+            expect_next_trigger_date_changed=False,
+        )
+
+        current_time = step_time(frozen_time, db_time, step_delta)
+        poll_scheduled_messages()
+        test_scheduled_message_attrs(
+            expected_next_trigger_date=current_time + delta,
+            expected_last_triggered_at=current_time,
+            expected_total_triggers=1,
+            expected_resolved=False,
+        )
+
+        current_time = step_time(frozen_time, db_time, step_delta)
+        poll_scheduled_messages()
+        # Since the scheduled message is be resolved, the expected_next_trigger_date will not be updated
+        test_scheduled_message_attrs(
+            # We subtract the offset here to account for it not being in the original `next_trigger_date`
+            expected_next_trigger_date=current_time - relativedelta(seconds=seconds_offset),
+            expected_last_triggered_at=current_time,
+            expected_total_triggers=2,
+            expected_resolved=True,
+            expect_next_trigger_date_changed=False,
+        )
+
+        # We are done now, but let's make 110% sure that another message will not be triggered
+        current_time = step_time(frozen_time, db_time, step_delta)
+        scheduled_message._trigger = Mock()
+        poll_scheduled_messages()
+        scheduled_message._trigger.assert_not_called()
+
+
+@pytest.mark.django_db()
+def test_error_when_sending_sending_message_to_a_user(caplog):
+    """This test makes sure that any error that happens when sending a message to a user does not affect other
+    pending messages"""
+
+    session = ExperimentSessionFactory()
+    experiment = session.experiment
+    team = experiment.team
+    schedule_conf = ScheduledMessageConfigFactory(
+        experiment=experiment,
+        team=team,
+        time_period=TimePeriod.DAYS,
+        frequency=1,
+        repetitions=2,
+    )
+    with (
+        caplog.at_level(logging.ERROR),
+        patch("apps.experiments.models.ExperimentSession.send_bot_message", side_effect=Exception("Oops")),
+        patch("apps.chat.tasks.functions.Now") as db_time,
+    ):
+        sm = ScheduledMessageFactory(participant=session.participant, schedule=schedule_conf)
+
+        # Let's put the DB time ahead of the scheduled message
+        utc_now = datetime.now().astimezone(pytz.timezone("UTC"))
+        db_time.return_value = utc_now + relativedelta(days=1.1)
+
+        pending_messages = _get_messages_to_fire()
+        assert len(pending_messages) == 1
+
+        poll_scheduled_messages()
+        assert len(caplog.records) == 1
+        for record in caplog.records:
+            record
```

Comment:
this doesn't seem to be doing anything

Reply:
:man_facepalming: right, I was busy checking logs and probably got distracted. Thankss :sweat_smile:

Reply:
https://github.com/dimagi/open-chat-studio/pull/373/commits/6996e196dc707b432776eb8dc8b6312948cd14a4